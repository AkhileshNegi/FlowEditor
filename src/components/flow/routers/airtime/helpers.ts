import {
    AirtimeRouterFormState,
    AirtimeTransferEntry
} from '~/components/flow/routers/airtime/AirtimeRouterForm';
import { createRenderNode } from '~/components/flow/routers/helpers';
import { Operators, Types } from '~/config/interfaces';
import {
    Case,
    Exit,
    RouterTypes,
    SwitchRouter,
    TransferAirtime,
    TransferAirtimeExitNames
} from '~/flowTypes';
import { RenderNode } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';
import { createUUID } from '~/utils';

export const nodeToState = (settings: NodeEditorSettings): AirtimeRouterFormState => {
    const originalAction = getOriginalAction(settings);

    const amounts: AirtimeTransferEntry[] = [];
    if (originalAction) {
        Object.keys(originalAction.amounts).forEach((key: string) => {
            amounts.push({ value: { code: key, amount: '' + originalAction.amounts[key] } });
        });
    }

    return {
        valid: false,
        amounts
    };
};

export const stateToNode = (
    settings: NodeEditorSettings,
    state: AirtimeRouterFormState
): RenderNode => {
    let uuid = createUUID();
    const originalAction = getOriginalAction(settings);
    if (originalAction) {
        uuid = originalAction.uuid;
    }

    const amounts = {};
    state.amounts.forEach((entry: AirtimeTransferEntry) => {
        if (entry.value.amount.trim().length > 0) {
            amounts[entry.value.code] = Number(entry.value.amount);
        }
    });

    const newAction: TransferAirtime = {
        uuid,
        type: Types.transfer_airtime,
        amounts
    };

    // If we're already a subflow, lean on those exits and cases
    let exits: Exit[];
    let cases: Case[];

    if (originalAction) {
        ({ exits } = settings.originalNode.node);
        ({ cases } = settings.originalNode.node.router as SwitchRouter);
    } else {
        // Otherwise, let's create some new ones
        exits = [
            {
                uuid: createUUID(),
                name: TransferAirtimeExitNames.Success,
                destination_node_uuid: null
            },
            {
                uuid: createUUID(),
                name: TransferAirtimeExitNames.Failure,
                destination_node_uuid: null
            }
        ];

        cases = [
            {
                uuid: createUUID(),
                type: Operators.has_webhook_status,
                arguments: ['success'],
                exit_uuid: exits[0].uuid
            },
            {
                uuid: createUUID(),
                type: Operators.has_webhook_status,
                arguments: ['failed'],
                exit_uuid: exits[1].uuid
            }
        ];
    }

    const router: SwitchRouter = {
        type: RouterTypes.switch,
        operand: '@child',
        cases,
        default_exit_uuid: null
    };

    const newRenderNode = createRenderNode(
        settings.originalNode.node.uuid,
        router,
        exits,
        Types.split_by_airtime,
        [newAction]
    );

    return newRenderNode;
};

export const getOriginalAction = (settings: NodeEditorSettings): TransferAirtime => {
    const action =
        settings.originalAction ||
        (settings.originalNode.node.actions.length > 0 && settings.originalNode.node.actions[0]);

    if (action.type === Types.transfer_airtime) {
        return action as TransferAirtime;
    }
};
