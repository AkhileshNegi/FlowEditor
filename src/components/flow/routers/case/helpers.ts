import { getOperatorConfig, Operator, Operators } from '~/config/operatorConfigs';
import {
    validate,
    validateLessThan,
    validateMoreThan,
    validateNumeric,
    validateRequired
} from '~/store/validators';
import { titleCase } from '~/utils';

import { CaseElementProps, CaseElementState } from './CaseElement';

export const initializeForm = (props: CaseElementProps): CaseElementState => {
    return {
        errors: [],
        operatorConfig: getOperatorConfig(props.kase.type),
        argument: {
            value:
                props.kase.arguments && props.kase.arguments.length === 1
                    ? props.kase.arguments[0]
                    : ''
        },
        min: {
            value:
                props.kase.arguments && props.kase.arguments.length === 2
                    ? props.kase.arguments[0]
                    : ''
        },
        max: {
            value:
                props.kase.arguments && props.kase.arguments.length === 2
                    ? props.kase.arguments[1]
                    : ''
        },
        exitName: { value: props.exitName || '' },
        exitNameEdited: !!props.exitName,
        valid: true
    };
};
/**
 * Determines prefix for case's exit name
 */
export const prefix = (operatorType: string): string => {
    let pre = '';

    if (operatorType.indexOf('_lt') > -1) {
        if (operatorType.indexOf('date') > -1) {
            pre = 'Before ';
        } else {
            if (operatorType.indexOf('lte') > -1) {
                pre = '<= ';
            } else {
                pre = '< ';
            }
        }
    } else if (operatorType.indexOf('_gt') > -1) {
        if (operatorType.indexOf('date') > -1) {
            pre = 'After ';
        } else {
            if (operatorType.indexOf('gte') > -1) {
                pre = '>= ';
            } else {
                pre = '>';
            }
        }
    }

    return pre;
};

/**
 * Returns min, max values for Operators.has_number_between case
 */
export const getMinMax = (args: string[] = []): { min: string; max: string } => {
    let min = '';
    let max = '';
    if (args.length) {
        if (strContainsNum(args[0])) {
            min = args[0];
        }
        if (args[1]) {
            if (strContainsNum(args[1])) {
                max = args[1];
            }
        }
    }
    return {
        min,
        max
    };
};

export const isFloat = (val: string): boolean => /^[+-]?\d?(\.\d*)?$/.test(val.trim());

export const isInt = (val: string): boolean => /^[\+\-]?\d+$/.test(val.trim());

export const strContainsNum = (str: string): boolean => {
    const trimmed = str.trim();
    if (isFloat(trimmed)) {
        return true;
    } else if (isInt(trimmed)) {
        return true;
    } else {
        return false;
    }
};

export const parseNum = (str: string): number => {
    const trimmed = (str || '').trim();
    if (isFloat(trimmed)) {
        return parseFloat(str);
    } else if (isInt(trimmed)) {
        return parseInt(trimmed, 10);
    }
};

export const validateCase = (keys: {
    operatorConfig: Operator;
    argument?: string;
    min?: string;
    max?: string;
    exitName?: string;
    exitEdited?: boolean;
}): Partial<CaseElementState> => {
    // when the exit is set, our arguments become required
    const validators = keys.exitEdited && keys.exitName ? [validateRequired] : [];

    const updates: Partial<CaseElementState> = {
        operatorConfig: keys.operatorConfig
    };

    if (keys.operatorConfig.operands > 0) {
        switch (keys.operatorConfig.type) {
            case Operators.has_number_eq:
            case Operators.has_number_gt:
            case Operators.has_number_gte:
            case Operators.has_number_lt:
            case Operators.has_number_lte:
                validators.push(validateNumeric);
                break;
        }

        if (keys.operatorConfig.type === Operators.has_number_between) {
            updates.min = validate(
                'Minimum value',
                keys.min || '',
                validators.concat([
                    validateNumeric,
                    validateLessThan(parseFloat(keys.max), 'the maximum')
                ])
            );

            updates.max = validate(
                'Maximum value',
                keys.max || '',
                validators.concat([
                    validateNumeric,
                    validateMoreThan(parseFloat(keys.min), 'the minimum')
                ])
            );

            updates.argument = { value: '' };
        } else {
            updates.min = { value: '' };
            updates.max = { value: '' };
            updates.argument = validate('Value', keys.argument || '', validators);
        }
    } else {
        // no operand clear them all
        updates.min = { value: '' };
        updates.max = { value: '' };
        updates.argument = { value: '' };
    }

    updates.exitNameEdited = !!keys.exitEdited;
    updates.exitName = { value: updates.exitNameEdited ? keys.exitName : getExitName(updates) };

    return updates;
};

export const getExitName = (state: Partial<CaseElementState>): string => {
    if (state.exitNameEdited) {
        return state.exitName.value;
    }

    if (state.operatorConfig.operands === 0) {
        return state.operatorConfig.categoryName;
    }

    if (
        state.operatorConfig.type === Operators.has_number_between &&
        state.min.value &&
        state.max.value
    ) {
        return `${state.min.value} - ${state.max.value}`;
    }

    if (state.argument && state.argument.value) {
        const pre = prefix(state.operatorConfig.type);
        const words = state.argument.value.match(/\w+/g);

        if (words && words.length > 0) {
            const [firstWord] = words;
            return pre + titleCase(firstWord);
        }

        return pre + titleCase(state.argument.value);
    }
    return '';
};
