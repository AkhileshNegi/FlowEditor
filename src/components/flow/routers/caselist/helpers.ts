import { DraggableStyle } from 'react-beautiful-dnd';
import { CaseProps, DragCursor } from '~/components/flow/routers/caselist/CaseList';
import { Operators } from '~/config/operatorConfigs';
import { Case, Exit } from '~/flowTypes';
import { createUUID } from '~/utils';

export const getExitName = (kase: Case, exits: Exit[]) => {
    const match = exits.find((exit: Exit) => exit.uuid === kase.exit_uuid);
    if (match) {
        return match.name;
    }
    return '';
};

export const createEmptyCase = (): CaseProps => {
    const uuid = createUUID();
    return {
        uuid,
        kase: {
            uuid,
            type: Operators.has_any_word,
            arguments: [''],
            exit_uuid: null
        },
        exitName: ''
    };
};

export const getItemStyle = (
    draggableStyle: DraggableStyle,
    isDragging: boolean
): DraggableStyle => ({
    userSelect: 'none',
    outline: 'none',
    ...draggableStyle,
    // Overwriting default draggableStyle object from this point down
    ...(isDragging
        ? {
              background: '#f2f9fc',
              borderRadius: 4,
              opacity: 0.75,
              height: draggableStyle.height + 15
          }
        : {})
});

export const getListStyle = (isDraggingOver: boolean, single: boolean): { cursor: DragCursor } => {
    if (single) {
        return null;
    }

    return {
        cursor: isDraggingOver ? DragCursor.move : DragCursor.pointer
    };
};
