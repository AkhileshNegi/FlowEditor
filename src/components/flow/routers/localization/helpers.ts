import { Types } from '~/config/typeConfigs';
import { Case, Exit, SwitchRouter } from '~/flowTypes';
import { LocalizedObject } from '~/services/Localization';
import { RenderNode } from '~/store/flowContext';
import { NodeEditorSettings } from '~/store/nodeEditor';

export enum LocalizedType {
    Exit,
    Case
}

export const getOriginal = (
    nodeSettings: NodeEditorSettings,
    localizedType: LocalizedType,
    uuid: string
) => {
    let items = nodeSettings.originalNode.node.exits;
    if (localizedType === LocalizedType.Case) {
        items = (nodeSettings.originalNode.node.router as SwitchRouter).cases;
    }
    return items.find((item: any) => item.uuid === uuid);
};

export const hasLocalizableCases = (renderNode: RenderNode) => {
    const type = renderNode.ui.type;
    return type === Types.wait_for_response || type === Types.split_by_expression;
};

export const getLocalizedObjects = (
    nodeSettings: NodeEditorSettings,
    localizedType: LocalizedType
): Exit[] | Case[] => {
    const filtered: any = [];

    let items: Exit[] | Case[] = nodeSettings.originalNode.node.exits;
    if (localizedType === LocalizedType.Case) {
        if (hasLocalizableCases(nodeSettings.originalNode)) {
            items = (nodeSettings.originalNode.node.router as SwitchRouter).cases;
        } else {
            items = [];
        }
    }

    for (const original of items) {
        const [localized] = nodeSettings.localizations.filter(
            (localizedObject: LocalizedObject) => localizedObject.getObject().uuid === original.uuid
        );

        if (localized.isLocalized()) {
            filtered.push(localized.getObject());
        } else {
            filtered.push({ uuid: original.uuid });
        }
    }
    return filtered;
};
