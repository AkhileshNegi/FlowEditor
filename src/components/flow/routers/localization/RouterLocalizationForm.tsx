import { react as bindCallbacks } from 'auto-bind';
import * as React from 'react';
import Dialog, { ButtonSet, HeaderStyle, Tab } from '~/components/dialog/Dialog';
import { determineTypeConfig } from '~/components/flow/helpers';
import { LocalizationFormProps } from '~/components/flow/props';
import {
    getLocalizedObjects,
    getOriginal,
    LocalizedType
} from '~/components/flow/routers/localization/helpers';
import * as styles from '~/components/flow/routers/localization/RouterLocalizationForm.scss';
import TextInputElement from '~/components/form/textinput/TextInputElement';
import { getOperatorConfig } from '~/config/operatorConfigs';
import { Case, Exit } from '~/flowTypes';
import { FormState, mergeForm } from '~/store/nodeEditor';

export interface RouterLocalizationFormState extends FormState {
    exits: Exit[];
    cases: Case[];
}

export default class RouterLocalizationForm extends React.Component<
    LocalizationFormProps,
    RouterLocalizationFormState
> {
    constructor(props: LocalizationFormProps) {
        super(props);

        const exits: Exit[] = getLocalizedObjects(props.nodeSettings, LocalizedType.Exit) as Exit[];
        const cases: Case[] = getLocalizedObjects(props.nodeSettings, LocalizedType.Case) as Case[];

        this.state = { exits, cases, valid: true };

        bindCallbacks(this, {
            include: [/^handle/]
        });
    }

    private handleUpdate(keys: { exit?: Exit; kase?: Case }): boolean {
        const updates: Partial<RouterLocalizationFormState> = {};

        if (keys.hasOwnProperty('exit')) {
            updates.exits = [keys.exit];
        }

        if (keys.hasOwnProperty('kase')) {
            updates.cases = [keys.kase];
        }

        const updated = mergeForm(this.state, updates);
        this.setState(updated);
        return updated.valid;
    }

    private handleUpdateExitName(exit: Exit, name: string): boolean {
        exit.name = name;
        return this.handleUpdate({ exit });
    }

    private handleUpdateCaseArgument(kase: Case, arg: string): boolean {
        kase.arguments = [arg];
        return this.handleUpdate({ kase });
    }

    private handleSave(): void {
        // collect up our exit localizations
        const translations: any[] = this.state.exits.map((exit: Exit) => {
            return exit.name
                ? {
                      uuid: exit.uuid,
                      translations: {
                          name: exit.name
                      }
                  }
                : { uuid: exit.uuid };
        });

        // same thing for any cases
        translations.push(
            ...this.state.cases.map((kase: Case) => {
                return kase.arguments
                    ? {
                          uuid: kase.uuid,
                          translations: {
                              arguments: kase.arguments
                          }
                      }
                    : { uuid: kase.uuid };
            })
        );

        this.props.updateLocalizations(this.props.language.id, translations);

        // notify our modal we are done
        this.props.onClose(false);
    }

    private getButtons(): ButtonSet {
        return {
            primary: { name: 'Ok', onClick: this.handleSave },
            secondary: { name: 'Cancel', onClick: () => this.props.onClose(true) }
        };
    }

    public renderCases(): JSX.Element[] {
        return this.state.cases.map((kase: Case) => {
            const originalCase = getOriginal(
                this.props.nodeSettings,
                LocalizedType.Case,
                kase.uuid
            ) as Case;

            const { verboseName } = getOperatorConfig(originalCase.type);

            const [orginalArgument] = originalCase.arguments;

            let argument = '';
            if (kase.arguments && kase.arguments.length > 0) {
                argument = kase.arguments[0];
            }

            return (
                <div
                    key={`translate_${kase.uuid}`}
                    data-spec="operator-field"
                    className={styles.translating_operator_container}
                >
                    <div data-spec="verbose-name" className={styles.translating_operator}>
                        {verboseName}
                    </div>
                    <div data-spec="argument-to-translate" className={styles.translating_from}>
                        {orginalArgument}
                    </div>
                    <div className={styles.translating_to}>
                        <TextInputElement
                            data-spec="localize-case"
                            name={kase.uuid}
                            placeholder={`${this.props.language.name} Translation`}
                            showLabel={false}
                            onChange={(arg: string) => this.handleUpdateCaseArgument(kase, arg)}
                            entry={{ value: argument }}
                        />
                    </div>
                </div>
            );
        });
    }

    public renderExits(): JSX.Element[] {
        return this.state.exits.map((exit: Exit) => {
            const originalExit = getOriginal(
                this.props.nodeSettings,
                LocalizedType.Exit,
                exit.uuid
            );

            const placeholder = `${this.props.language.name} Translation`;

            if (!exit.name) {
                exit.name = '';
            }

            return (
                <div key={exit.uuid} className={styles.translating_exit}>
                    <div data-spec="exit-name" className={styles.translating_from}>
                        {originalExit.name}
                    </div>
                    <div className={styles.translating_to}>
                        <TextInputElement
                            data-spec="localize-exit"
                            name={exit.name}
                            placeholder={placeholder}
                            showLabel={false}
                            entry={{ value: exit.name }}
                            onChange={(name: string) => this.handleUpdateExitName(exit, name)}
                        />
                    </div>
                </div>
            );
        });
    }

    public render(): JSX.Element {
        const typeConfig = determineTypeConfig(this.props.nodeSettings);

        const tabs: Tab[] = [];

        if (this.state.cases.length > 0) {
            tabs.push({
                name: 'Rule Translations',
                body: (
                    <>
                        <p data-spec="instructions">
                            Sometimes languages need special rules to route things properly. If a
                            translation is not provided, the original rule will be used.
                        </p>
                        {this.renderCases()}
                    </>
                )
            });
        }

        const exits = (
            <Dialog
                title={`${this.props.language.name} Category Names`}
                headerClass={typeConfig.type}
                buttons={this.getButtons()}
                tabs={tabs}
            >
                <p data-spec="instructions">
                    When category names are referenced later in the flow, the appropriate language
                    for the category will be used. If no translation is provided, the original text
                    will be used.
                </p>
                {this.renderExits()}
            </Dialog>
        );

        const back = (
            <Dialog
                title={`${this.props.language.name} Rules`}
                headerStyle={HeaderStyle.BARBER}
                headerClass={typeConfig.type}
                headerIcon="fe-cog"
                subtitle="Advanced Settings"
                buttons={this.getButtons()}
            >
                <p data-spec="instructions">
                    Sometimes languages need special rules to route things properly. If a
                    translation is not provided, the original rule will be used.
                </p>
                {this.renderCases()}
            </Dialog>
        );

        // if we have cases, use a flipper

        return exits;
    }
}
