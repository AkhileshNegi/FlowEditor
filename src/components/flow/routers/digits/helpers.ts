import { CaseProps } from '~/components/flow/routers/caselist/CaseList';
import {
    createCaseProps,
    createRenderNode,
    hasCases,
    resolveExits
} from '~/components/flow/routers/helpers';
import { Types } from '~/config/interfaces';
import { getType } from '~/config/typeConfigs';
import { HintTypes, Router, RouterTypes, SwitchRouter, WaitTypes } from '~/flowTypes';
import { RenderNode } from '~/store/flowContext';
import { NodeEditorSettings, StringEntry } from '~/store/nodeEditor';

import { DigitsRouterFormState } from './DigitsRouterForm';

export const nodeToState = (settings: NodeEditorSettings): DigitsRouterFormState => {
    let initialCases: CaseProps[] = [];

    // TODO: work out an incremental result name
    let resultName: StringEntry = { value: '' };

    if (getType(settings.originalNode) === Types.wait_for_digits) {
        const router = settings.originalNode.node.router as SwitchRouter;
        if (router) {
            if (hasCases(settings.originalNode.node)) {
                initialCases = createCaseProps(router.cases, settings.originalNode.node.exits);
            }

            resultName = { value: router.result_name || '' };
        }
    }

    return {
        cases: initialCases,
        resultName,
        valid: true
    };
};

export const stateToNode = (
    settings: NodeEditorSettings,
    state: DigitsRouterFormState
): RenderNode => {
    const { cases, exits, defaultExit } = resolveExits(state.cases, false, settings);

    const optionalRouter: Pick<Router, 'result_name'> = {};
    if (state.resultName.value) {
        optionalRouter.result_name = state.resultName.value;
    }

    const router: SwitchRouter = {
        type: RouterTypes.switch,
        default_exit_uuid: defaultExit,
        cases,
        operand: '@input',
        ...optionalRouter
    };

    const newRenderNode = createRenderNode(
        settings.originalNode.node.uuid,
        router,
        exits,
        Types.wait_for_response,
        [],
        { type: WaitTypes.msg, hint: { type: HintTypes.digits } }
    );

    return newRenderNode;
};
