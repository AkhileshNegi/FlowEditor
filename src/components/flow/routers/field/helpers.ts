import { NAME_PROPERTY } from '~/components/flow/props';
import { CaseProps } from '~/components/flow/routers/caselist/CaseList';
import { FieldRouterFormState } from '~/components/flow/routers/field/FieldRouterForm';
import {
    createCaseProps,
    createRenderNode,
    hasCases,
    resolveExits
} from '~/components/flow/routers/helpers';
import { DEFAULT_OPERAND } from '~/components/nodeeditor/constants';
import { Types } from '~/config/typeConfigs';
import { Router, RouterTypes, SwitchRouter } from '~/flowTypes';
import { AssetType, RenderNode } from '~/store/flowContext';
import { NodeEditorSettings, StringEntry } from '~/store/nodeEditor';

export const nodeToState = (settings: NodeEditorSettings): FieldRouterFormState => {
    let initialCases: CaseProps[] = [];

    // TODO: work out an incremental result name
    let resultName: StringEntry = { value: '' };

    let field: any = NAME_PROPERTY;

    if (settings.originalNode && settings.originalNode.ui.type === Types.split_by_contact_field) {
        const router = settings.originalNode.node.router as SwitchRouter;

        if (router) {
            if (hasCases(settings.originalNode.node)) {
                initialCases = createCaseProps(router.cases, settings.originalNode.node.exits);
            }

            resultName = { value: router.result_name || '' };
        }

        const operand = settings.originalNode.ui.config.operand;
        field = { id: operand.id, type: operand.type, name: operand.name };
    }

    return {
        cases: initialCases,
        resultName,
        field: { value: field },
        valid: true
    };
};

export const stateToNode = (
    settings: NodeEditorSettings,
    state: FieldRouterFormState
): RenderNode => {
    const { cases, exits, defaultExit } = resolveExits(state.cases, false, settings);

    const optionalRouter: Pick<Router, 'result_name'> = {};
    if (state.resultName.value) {
        optionalRouter.result_name = state.resultName.value;
    }

    let operand = DEFAULT_OPERAND;
    const asset = state.field.value;
    if (asset.type === AssetType.URN) {
        operand = `@(format_urn(contact.urns.${asset.id}))`;
    } else if (asset.type === AssetType.Field) {
        operand = `@contact.fields.${asset.id}`;
    } else {
        operand = `@contact.${asset.id}`;
    }

    const router: SwitchRouter = {
        type: RouterTypes.switch,
        default_exit_uuid: defaultExit,
        cases,
        operand,
        ...optionalRouter
    };

    const newRenderNode = createRenderNode(
        settings.originalNode.node.uuid,
        router,
        exits,
        Types.split_by_contact_field,
        [],
        null,
        {
            operand: {
                id: asset.id,
                type: asset.type,
                name: asset.name
            }
        }
    );

    return newRenderNode;
};
