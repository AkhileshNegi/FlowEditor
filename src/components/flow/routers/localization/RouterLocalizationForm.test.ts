import { LocalizationFormProps } from '~/components/flow/props';
import RouterLocalizationForm from '~/components/flow/routers/localization/RouterLocalizationForm';
import { DEFAULT_OPERAND } from '~/components/nodeeditor/constants';
import { Operators } from '~/config/interfaces';
import { RouterTypes, SwitchRouter } from '~/flowTypes';
import { getLocalizations } from '~/store/helpers';
import { composeComponentTestUtils, mock } from '~/testUtils';
import { createRenderNode, Spanish } from '~/testUtils/assetCreators';
import * as utils from '~/utils';
import { createUUID } from '~/utils';

mock(utils, 'createUUID', utils.seededUUIDs());

const colorsExit = createUUID();
const otherExit = createUUID();
const redCase = createUUID();
const blueCase = createUUID();

const responseRenderNode = createRenderNode({
    actions: [],
    exits: [
        { destination_node_uuid: null, name: 'Colors', uuid: colorsExit },
        { destination_node_uuid: null, name: 'Other', uuid: otherExit }
    ],
    router: {
        type: RouterTypes.switch,
        operand: DEFAULT_OPERAND,
        cases: [
            {
                uuid: redCase,
                type: Operators.has_any_word,
                arguments: ['red'],
                exit_uuid: colorsExit
            },
            {
                uuid: blueCase,
                type: Operators.has_any_word,
                arguments: ['blue'],
                exit_uuid: colorsExit
            }
        ],
        default_exit_uuid: otherExit,
        result_name: 'Color'
    } as SwitchRouter
});

const localizations = getLocalizations(responseRenderNode.node, null, Spanish, {
    [otherExit]: { name: ['Otro'] },
    [redCase]: { arguments: ['rojo, r'] }
});

const baseProps: LocalizationFormProps = {
    language: Spanish,
    updateLocalizations: jest.fn(),
    onClose: jest.fn(),
    nodeSettings: {
        originalNode: responseRenderNode,
        originalAction: null,
        localizations
    }
};

const { setup } = composeComponentTestUtils<LocalizationFormProps>(
    RouterLocalizationForm,
    baseProps
);

describe(RouterLocalizationForm.name, () => {
    it('should render', () => {
        const { wrapper } = setup(true);
        expect(wrapper).toMatchSnapshot();
    });

    it('initializes', () => {
        const { wrapper } = setup(true);
        expect(wrapper).toMatchSnapshot();
    });

    describe('updates', () => {
        it('should save changes', () => {
            const { instance, props } = setup(true);

            instance.handleUpdateExitName(responseRenderNode.node.exits[0], 'Roooojo!');
            instance.handleUpdateCaseArgument(
                (responseRenderNode.node.router as SwitchRouter).cases[0],
                'Red, r, rolo, maroon'
            );
            expect(instance.state).toMatchSnapshot();

            instance.handleSave();
            expect(props.updateLocalizations).toHaveBeenCalled();
            expect(props.updateLocalizations).toMatchCallSnapshot();
        });
    });
});
