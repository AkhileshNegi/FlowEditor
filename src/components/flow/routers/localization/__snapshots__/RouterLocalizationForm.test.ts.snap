// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouterLocalizationForm initializes 1`] = `
<Flipper
  back={
    <Dialog
      buttons={
        Object {
          "primary": Object {
            "name": "Ok",
            "onClick": [Function],
          },
          "secondary": Object {
            "name": "Cancel",
            "onClick": [Function],
          },
        }
      }
      headerClass="split_by_expression"
      headerIcon="fe-cog"
      headerStyle="barber"
      subtitle="Advanced Settings"
      title="Spanish Rules"
    >
      <p
        data-spec="instructions"
      >
        Sometimes languages need special rules to route things properly. If a translation is not provided, the original rule will be used.
      </p>
      <div
        className="translating_operator_container"
        data-spec="operator-field"
      >
        <div
          className="translating_operator"
          data-spec="verbose-name"
        >
          has any of the words
        </div>
        <div
          className="translating_from"
          data-spec="argument-to-translate"
        >
          red
        </div>
        <div
          className="translating_to"
        >
          <Connect(TextInputElement)
            data-spec="localize-case"
            entry={
              Object {
                "value": "rojo, r",
              }
            }
            name="generated_uuid_3"
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
      <div
        className="translating_operator_container"
        data-spec="operator-field"
      >
        <div
          className="translating_operator"
          data-spec="verbose-name"
        >
          has any of the words
        </div>
        <div
          className="translating_from"
          data-spec="argument-to-translate"
        >
          blue
        </div>
        <div
          className="translating_to"
        >
          <Connect(TextInputElement)
            data-spec="localize-case"
            entry={
              Object {
                "value": "",
              }
            }
            name="generated_uuid_4"
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
    </Dialog>
  }
  front={
    <Dialog
      buttons={
        Object {
          "primary": Object {
            "name": "Ok",
            "onClick": [Function],
          },
          "secondary": Object {
            "name": "Cancel",
            "onClick": [Function],
          },
        }
      }
      headerClass="split_by_expression"
      title="Spanish Category Names"
    >
      <p
        data-spec="instructions"
      >
        When category names are referenced later in the flow, the appropriate language for the category will be used. If no translation is provided, the original text will be used.
      </p>
      <div
        className="translating_exit"
      >
        <div
          className="translating_from"
          data-spec="exit-name"
        >
          Colors
        </div>
        <div
          className="translating_to"
        >
          <Connect(TextInputElement)
            data-spec="localize-exit"
            entry={
              Object {
                "value": "",
              }
            }
            name=""
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
      <div
        className="translating_exit"
      >
        <div
          className="translating_from"
          data-spec="exit-name"
        >
          Other
        </div>
        <div
          className="translating_to"
        >
          <Connect(TextInputElement)
            data-spec="localize-exit"
            entry={
              Object {
                "value": Array [
                  "Otro",
                ],
              }
            }
            name={
              Array [
                "Otro",
              ]
            }
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
    </Dialog>
  }
/>
`;

exports[`RouterLocalizationForm should render 1`] = `
<Flipper
  back={
    <Dialog
      buttons={
        Object {
          "primary": Object {
            "name": "Ok",
            "onClick": [Function],
          },
          "secondary": Object {
            "name": "Cancel",
            "onClick": [Function],
          },
        }
      }
      headerClass="split_by_expression"
      headerIcon="fe-cog"
      headerStyle="barber"
      subtitle="Advanced Settings"
      title="Spanish Rules"
    >
      <p
        data-spec="instructions"
      >
        Sometimes languages need special rules to route things properly. If a translation is not provided, the original rule will be used.
      </p>
      <div
        className="translating_operator_container"
        data-spec="operator-field"
      >
        <div
          className="translating_operator"
          data-spec="verbose-name"
        >
          has any of the words
        </div>
        <div
          className="translating_from"
          data-spec="argument-to-translate"
        >
          red
        </div>
        <div
          className="translating_to"
        >
          <Connect(TextInputElement)
            data-spec="localize-case"
            entry={
              Object {
                "value": "rojo, r",
              }
            }
            name="generated_uuid_3"
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
      <div
        className="translating_operator_container"
        data-spec="operator-field"
      >
        <div
          className="translating_operator"
          data-spec="verbose-name"
        >
          has any of the words
        </div>
        <div
          className="translating_from"
          data-spec="argument-to-translate"
        >
          blue
        </div>
        <div
          className="translating_to"
        >
          <Connect(TextInputElement)
            data-spec="localize-case"
            entry={
              Object {
                "value": "",
              }
            }
            name="generated_uuid_4"
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
    </Dialog>
  }
  front={
    <Dialog
      buttons={
        Object {
          "primary": Object {
            "name": "Ok",
            "onClick": [Function],
          },
          "secondary": Object {
            "name": "Cancel",
            "onClick": [Function],
          },
        }
      }
      headerClass="split_by_expression"
      title="Spanish Category Names"
    >
      <p
        data-spec="instructions"
      >
        When category names are referenced later in the flow, the appropriate language for the category will be used. If no translation is provided, the original text will be used.
      </p>
      <div
        className="translating_exit"
      >
        <div
          className="translating_from"
          data-spec="exit-name"
        >
          Colors
        </div>
        <div
          className="translating_to"
        >
          <Connect(TextInputElement)
            data-spec="localize-exit"
            entry={
              Object {
                "value": "",
              }
            }
            name=""
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
      <div
        className="translating_exit"
      >
        <div
          className="translating_from"
          data-spec="exit-name"
        >
          Other
        </div>
        <div
          className="translating_to"
        >
          <Connect(TextInputElement)
            data-spec="localize-exit"
            entry={
              Object {
                "value": Array [
                  "Otro",
                ],
              }
            }
            name={
              Array [
                "Otro",
              ]
            }
            onChange={[Function]}
            placeholder="Spanish Translation"
            showLabel={false}
          />
        </div>
      </div>
    </Dialog>
  }
/>
`;

exports[`RouterLocalizationForm updates should save changes 1`] = `
Object {
  "cases": Array [
    Object {
      "arguments": Array [
        "Red, r, rolo, maroon",
      ],
      "exit_uuid": "generated_uuid_1",
      "type": "has_any_word",
      "uuid": "generated_uuid_3",
    },
    Object {
      "uuid": "generated_uuid_4",
    },
  ],
  "exits": Array [
    Object {
      "destination_node_uuid": null,
      "name": "Roooojo!",
      "uuid": "generated_uuid_1",
    },
    Object {
      "destination_node_uuid": null,
      "name": Array [
        "Otro",
      ],
      "uuid": "generated_uuid_2",
    },
  ],
  "valid": true,
}
`;

exports[`RouterLocalizationForm updates should save changes 2`] = `
Array [
  "spa",
  Array [
    Object {
      "translations": Object {
        "name": "Roooojo!",
      },
      "uuid": "generated_uuid_1",
    },
    Object {
      "translations": Object {
        "name": Array [
          "Otro",
        ],
      },
      "uuid": "generated_uuid_2",
    },
    Object {
      "translations": Object {
        "arguments": Array [
          "Red, r, rolo, maroon",
        ],
      },
      "uuid": "generated_uuid_3",
    },
    Object {
      "uuid": "generated_uuid_4",
    },
  ],
]
`;
